File: sdks/go/pkg/klikkflow/executions.go
package reporunner

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"github.com/gorilla/websocket"
	"golang.org/x/sync"
)

// ExecutionManager handles all execution-related operations
type ExecutionManager struct {
	client              *Client
	executionListeners  map[string][]func(*ExecutionData)
	listenerMutex       sync.RWMutex
	wsConnection        *websocket.Conn
	wsContext           context.Context
	wsCancel            context.CancelFunc
	wsConnected         bool
	wsMutex             sync.RWMutex
}

// NewExecutionManager creates a new execution manager
func NewExecutionManager(client *Client) *ExecutionManager {
	return &ExecutionManager{
		client:             client,
		executionListeners: make(map[string][]func(*ExecutionData)),
	}
}

// Start starts a new workflow execution
func (em *ExecutionManager) Start(workflowID string, inputData map[string]interface{}, options *ExecutionOptions) (*Execution, error) {
	if workflowID == "" {
		return nil, &ValidationError{Message: "workflow ID is required"}
	}

	executeReq := map[string]interface{}{
		"workflowId": workflowID,
		"inputData":  inputData,
	}

	// Add optional parameters
	if options != nil {
		if options.ExecutionMode != "" {
			executeReq["executionMode"] = options.ExecutionMode
		}
		if options.WebhookURL != "" {
			executeReq["webhookUrl"] = options.WebhookURL
		}
		if options.Metadata != nil {
			executeReq["metadata"] = options.Metadata
		}
		if options.StartFrom != "" {
			executeReq["startFrom"] = options.StartFrom
		}
		if options.TestMode {
			executeReq["mode"] = "test"
		}
	}

	var response APIResponse[Execution]
	if err := em.client.makeRequest(http.MethodPost, "/api/executions", executeReq, &response); err != nil {
		return nil, fmt.Errorf("failed to start execution: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to start execution: %s", response.Error)
	}

	return &response.Data, nil
}

// Get retrieves a specific execution by ID
func (em *ExecutionManager) Get(executionID string) (*Execution, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	var response APIResponse[Execution]
	path := fmt.Sprintf("/api/executions/%s", url.PathEscape(executionID))

	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to get execution: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to get execution: %s", response.Error)
	}

	return &response.Data, nil
}

// List retrieves executions with optional filtering and pagination
func (em *ExecutionManager) List(filter *ExecutionFilter, pagination *PaginationParams) (*PaginatedResponse[Execution], error) {
	params := url.Values{}

	// Add filter parameters
	if filter != nil {
		if filter.WorkflowID != nil {
			params.Add("workflowId", *filter.WorkflowID)
		}
		if filter.Status != nil {
			params.Add("status", string(*filter.Status))
		}
		if filter.StartDate != nil {
			params.Add("startDate", filter.StartDate.Format(time.RFC3339))
		}
		if filter.EndDate != nil {
			params.Add("endDate", filter.EndDate.Format(time.RFC3339))
		}
		if filter.Mode != nil {
			params.Add("mode", *filter.Mode)
		}
		for _, tag := range filter.Tags {
			params.Add("tags", tag)
		}
	}

	// Add pagination parameters
	if pagination != nil {
		if pagination.Page > 0 {
			params.Add("page", strconv.Itoa(pagination.Page))
		}
		if pagination.PageSize > 0 {
			params.Add("pageSize", strconv.Itoa(pagination.PageSize))
		}
		if pagination.SortBy != "" {
			params.Add("sortBy", pagination.SortBy)
		}
		if pagination.SortDir != "" {
			params.Add("sortDir", pagination.SortDir)
		}
	}

	path := "/api/executions"
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var response APIResponse[PaginatedResponse[Execution]]
	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to list executions: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to list executions: %s", response.Error)
	}

	return &response.Data, nil
}

// Cancel cancels a running execution
func (em *ExecutionManager) Cancel(executionID string) error {
	if executionID == "" {
		return &ValidationError{Message: "execution ID is required"}
	}

	var response APIResponse[map[string]interface{}]
	path := fmt.Sprintf("/api/executions/%s/cancel", url.PathEscape(executionID))

	if err := em.client.makeRequest(http.MethodPost, path, nil, &response); err != nil {
		return fmt.Errorf("failed to cancel execution: %w", err)
	}

	if !response.Success {
		return fmt.Errorf("failed to cancel execution: %s", response.Error)
	}

	return nil
}

// Retry retries a failed execution, optionally from a specific node
func (em *ExecutionManager) Retry(executionID string, fromNode *string) (*Execution, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	retryReq := map[string]interface{}{}
	if fromNode != nil {
		retryReq["fromNode"] = *fromNode
	}

	var response APIResponse[Execution]
	path := fmt.Sprintf("/api/executions/%s/retry", url.PathEscape(executionID))

	if err := em.client.makeRequest(http.MethodPost, path, retryReq, &response); err != nil {
		return nil, fmt.Errorf("failed to retry execution: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to retry execution: %s", response.Error)
	}

	return &response.Data, nil
}

// Pause pauses a running execution
func (em *ExecutionManager) Pause(executionID string) error {
	if executionID == "" {
		return &ValidationError{Message: "execution ID is required"}
	}

	var response APIResponse[interface{}]
	path := fmt.Sprintf("/api/executions/%s/pause", url.PathEscape(executionID))

	if err := em.client.makeRequest(http.MethodPost, path, nil, &response); err != nil {
		return fmt.Errorf("failed to pause execution: %w", err)
	}

	if !response.Success {
		return fmt.Errorf("failed to pause execution: %s", response.Error)
	}

	return nil
}

// Resume resumes a paused execution
func (em *ExecutionManager) Resume(executionID string) error {
	if executionID == "" {
		return &ValidationError{Message: "execution ID is required"}
	}

	var response APIResponse[interface{}]
	path := fmt.Sprintf("/api/executions/%s/resume", url.PathEscape(executionID))

	if err := em.client.makeRequest(http.MethodPost, path, nil, &response); err != nil {
		return fmt.Errorf("failed to resume execution: %w", err)
	}

	if !response.Success {
		return fmt.Errorf("failed to resume execution: %s", response.Error)
	}

	return nil
}

// GetLogs retrieves logs for an execution
func (em *ExecutionManager) GetLogs(executionID string, nodeID *string, level *string) ([]map[string]interface{}, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	params := url.Values{}
	if nodeID != nil {
		params.Add("nodeId", *nodeID)
	}
	if level != nil {
		params.Add("level", *level)
	}

	path := fmt.Sprintf("/api/executions/%s/logs", url.PathEscape(executionID))
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var response APIResponse[[]map[string]interface{}]
	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to get execution logs: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to get execution logs: %s", response.Error)
	}

	return response.Data, nil
}

// GetNodeData retrieves execution data for a specific node
func (em *ExecutionManager) GetNodeData(executionID string, nodeID string) (*ExecutionData, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	if nodeID == "" {
		return nil, &ValidationError{Message: "node ID is required"}
	}

	var response APIResponse[ExecutionData]
	path := fmt.Sprintf("/api/executions/%s/nodes/%s", url.PathEscape(executionID), url.PathEscape(nodeID))

	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to get node execution data: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to get node execution data: %s", response.Error)
	}

	return &response.Data, nil
}

// WaitForCompletion waits for an execution to complete with optional timeout
func (em *ExecutionManager) WaitForCompletion(executionID string, timeout *time.Duration) (*Execution, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	start := time.Now()
	pollInterval := 2 * time.Second

	for {
		execution, err := em.Get(executionID)
		if err != nil {
			return nil, err
		}

		// Check if execution is complete
		if execution.Status == ExecutionStatusCompleted ||
			execution.Status == ExecutionStatusFailed ||
			execution.Status == ExecutionStatusCancelled {
			return execution, nil
		}

		// Check timeout
		if timeout != nil && time.Since(start) > *timeout {
			return nil, &TimeoutError{
				Message:   fmt.Sprintf("execution %s did not complete within %v", executionID, *timeout),
				Operation: "wait_for_completion",
				Timeout:   timeout.String(),
			}
		}

		// Wait before next poll
		time.Sleep(pollInterval)
	}
}

// StreamUpdates streams real-time updates for a running execution via WebSocket
func (em *ExecutionManager) StreamUpdates(ctx context.Context, executionID string, callback func(*ExecutionData)) error {
	if executionID == "" {
		return &ValidationError{Message: "execution ID is required"}
	}

	if callback == nil {
		return &ValidationError{Message: "callback function is required"}
	}

	// Add listener
	em.AddListener(executionID, callback)
	defer em.RemoveListener(executionID, callback)

	// Connect to WebSocket if not already connected
	if err := em.connectWebSocket(ctx); err != nil {
		return fmt.Errorf("failed to connect to WebSocket: %w", err)
	}

	// Subscribe to execution updates
	subscribeMsg := map[string]interface{}{
		"type":        "subscribe",
		"executionId": executionID,
	}

	if err := em.sendWebSocketMessage(subscribeMsg); err != nil {
		return fmt.Errorf("failed to subscribe to execution updates: %w", err)
	}

	// Wait for context cancellation
	<-ctx.Done()

	// Unsubscribe from execution updates
	unsubscribeMsg := map[string]interface{}{
		"type":        "unsubscribe",
		"executionId": executionID,
	}

	if err := em.sendWebSocketMessage(unsubscribeMsg); err != nil {
		em.client.logger.Warn().Err(err).Msg("Failed to unsubscribe from execution updates")
	}

	return ctx.Err()
}

// AddListener adds a callback function for execution updates
func (em *ExecutionManager) AddListener(executionID string, callback func(*ExecutionData)) {
	em.listenerMutex.Lock()
	defer em.listenerMutex.Unlock()

	if em.executionListeners[executionID] == nil {
		em.executionListeners[executionID] = make([]func(*ExecutionData), 0)
	}

	em.executionListeners[executionID] = append(em.executionListeners[executionID], callback)
}

// RemoveListener removes a callback function for execution updates
func (em *ExecutionManager) RemoveListener(executionID string, callback func(*ExecutionData)) {
	em.listenerMutex.Lock()
	defer em.listenerMutex.Unlock()

	listeners := em.executionListeners[executionID]
	for i, listener := range listeners {
		// Compare function pointers (this is a simplified comparison)
		if fmt.Sprintf("%p", listener) == fmt.Sprintf("%p", callback) {
			em.executionListeners[executionID] = append(listeners[:i], listeners[i+1:]...)
			break
		}
	}

	// Clean up empty listener lists
	if len(em.executionListeners[executionID]) == 0 {
		delete(em.executionListeners, executionID)
	}
}

// ExportData exports execution data in various formats
func (em *ExecutionManager) ExportData(executionID string, format string) ([]byte, error) {
	if executionID == "" {
		return nil, &ValidationError{Message: "execution ID is required"}
	}

	if format == "" {
		format = "json"
	}

	params := url.Values{}
	params.Add("format", format)

	path := fmt.Sprintf("/api/executions/%s/export?%s", url.PathEscape(executionID), params.Encode())

	var response APIResponse[map[string]interface{}]
	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to export execution data: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to export execution data: %s", response.Error)
	}

	// Convert to appropriate format
	if format == "json" {
		return json.Marshal(response.Data)
	}

	// For other formats, assume the API returns binary data
	if data, ok := response.Data["data"].(string); ok {
		return []byte(data), nil
	}

	return nil, fmt.Errorf("unexpected response format")
}

// GetStatistics retrieves execution statistics
func (em *ExecutionManager) GetStatistics(workflowID *string, timeRange map[string]time.Time) (*ExecutionStatistics, error) {
	params := url.Values{}

	if workflowID != nil {
		params.Add("workflowId", *workflowID)
	}

	if timeRange != nil {
		if start, ok := timeRange["start"]; ok {
			params.Add("startDate", start.Format(time.RFC3339))
		}
		if end, ok := timeRange["end"]; ok {
			params.Add("endDate", end.Format(time.RFC3339))
		}
	}

	path := "/api/executions/statistics"
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var response APIResponse[ExecutionStatistics]
	if err := em.client.makeRequest(http.MethodGet, path, nil, &response); err != nil {
		return nil, fmt.Errorf("failed to get execution statistics: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("failed to get execution statistics: %s", response.Error)
	}

	return &response.Data, nil
}

// connectWebSocket establishes a WebSocket connection
func (em *ExecutionManager) connectWebSocket(ctx context.Context) error {
	em.wsMutex.Lock()
	defer em.wsMutex.Unlock()

	if em.wsConnected {
		return nil
	}

	// Get WebSocket URL from config
	wsConfig := em.client.config.WebSocketConfig
	if wsConfig == nil || wsConfig.URL == "" {
		return &WebSocketError{Message: "WebSocket URL not configured"}
	}

	// Create WebSocket connection
	header := http.Header{}
	if token := em.client.authManager.GetCurrentToken(); token != nil {
		header.Set("Authorization", "Bearer "+token.AccessToken)
	}

	conn, _, err := websocket.DefaultDialer.Dial(wsConfig.URL, header)
	if err != nil {
		return &WebSocketError{
			Message: fmt.Sprintf("failed to connect to WebSocket: %v", err),
		}
	}

	em.wsConnection = conn
	em.wsContext, em.wsCancel = context.WithCancel(ctx)
	em.wsConnected = true

	// Start message handling goroutine
	go em.handleWebSocketMessages()

	em.client.logger.Debug().Msg("WebSocket connection established")
	return nil
}

// sendWebSocketMessage sends a message via WebSocket
func (em *ExecutionManager) sendWebSocketMessage(message map[string]interface{}) error {
	em.wsMutex.RLock()
	defer em.wsMutex.RUnlock()

	if !em.wsConnected || em.wsConnection == nil {
		return &WebSocketError{Message: "WebSocket not connected"}
	}

	return em.wsConnection.WriteJSON(message)
}

// handleWebSocketMessages handles incoming WebSocket messages
func (em *ExecutionManager) handleWebSocketMessages() {
	defer func() {
		em.wsMutex.Lock()
		em.wsConnected = false
		if em.wsConnection != nil {
			em.wsConnection.Close()
			em.wsConnection = nil
		}
		if em.wsCancel != nil {
			em.wsCancel()
		}
		em.wsMutex.Unlock()
	}()

	for {
		select {
		case <-em.wsContext.Done():
			return
		default:
			var message WebSocketMessage
			if err := em.wsConnection.ReadJSON(&message); err != nil {
				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
					em.client.logger.Error().Err(err).Msg("WebSocket error")
				}
				return
			}

			// Process the message
			em.processWebSocketMessage(&message)
		}
	}
}

// processWebSocketMessage processes a WebSocket message
func (em *ExecutionManager) processWebSocketMessage(message *WebSocketMessage) {
	if message.Type != "execution_update" || message.ExecutionID == "" {
		return
	}

	// Convert message data to ExecutionData
	var executionData ExecutionData
	if dataBytes, err := json.Marshal(message.Data); err == nil {
		if err := json.Unmarshal(dataBytes, &executionData); err != nil {
			em.client.logger.Warn().Err(err).Msg("Failed to unmarshal execution data")
			return
		}
	}

	// Notify listeners
	em.listenerMutex.RLock()
	listeners := em.executionListeners[message.ExecutionID]
	em.listenerMutex.RUnlock()

	for _, listener := range listeners {
		go listener(&executionData)
	}
}

// ExecutionOptions represents options for starting an execution
type ExecutionOptions struct {
	ExecutionMode string                 `json:"executionMode,omitempty"`
	WebhookURL    string                 `json:"webhookUrl,omitempty"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
	StartFrom     string                 `json:"startFrom,omitempty"`
	TestMode      bool                   `json:"testMode,omitempty"`
}

// Close closes the execution manager and cleans up resources
func (em *ExecutionManager) Close() error {
	em.wsMutex.Lock()
	defer em.wsMutex.Unlock()

	if em.wsCancel != nil {
		em.wsCancel()
	}

	if em.wsConnection != nil {
		em.wsConnection.Close()
		em.wsConnection = nil
	}

	em.wsConnected = false

	em.listenerMutex.Lock()
	em.executionListeners = make(map[string][]func(*ExecutionData))
	em.listenerMutex.Unlock()

	return nil
}

File: sdks/python/reporunner/executions.py
"""
Execution management for the KlikkFlow Python SDK.
This module provides comprehensive execution operations including real-time monitoring,
cancellation, and detailed execution data retrieval.
"""

from typing import Dict, List, Optional, Any, AsyncGenerator, Callable
from datetime import datetime
import asyncio
import json

from .types import (
    Execution, 
    ExecutionStatus, 
    ExecutionFilter,
    ExecutionNode,
    ExecutionData,
    PaginationParams,
    PaginatedResponse,
    WebSocketMessage
)
from .exceptions import (
    KlikkFlowAPIError,
    ExecutionNotFoundError,
    ExecutionError,
    ValidationError
)


class ExecutionManager:
    """
    Manages workflow execution operations including monitoring, cancellation, and data retrieval.
    
    This class provides methods for:
    - Starting and managing workflow executions
    - Real-time execution monitoring via WebSocket
    - Retrieving execution history and details
    - Cancelling running executions
    """
    
    def __init__(self, client: Any):
        """Initialize the execution manager with a client instance."""
        self.client = client
        self._execution_listeners: Dict[str, List[Callable]] = {}
        self._websocket_connection = None

    async def start_execution(
        self,
        workflow_id: str,
        input_data: Optional[Dict[str, Any]] = None,
        execution_mode: str = "full",
        webhook_url: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Execution:
        """
        Start a new workflow execution.
        
        Args:
            workflow_id: ID of the workflow to execute
            input_data: Initial input data for the workflow
            execution_mode: Execution mode ('full', 'test', 'debug')
            webhook_url: Optional webhook URL for completion notifications
            metadata: Optional metadata to attach to the execution
            
        Returns:
            Execution object with initial execution details
            
        Raises:
            ValidationError: If workflow_id or parameters are invalid
            KlikkFlowAPIError: If API request fails
        """
        if not workflow_id:
            raise ValidationError("workflow_id is required")
            
        payload = {
            "workflowId": workflow_id,
            "executionMode": execution_mode,
            "inputData": input_data or {},
            "webhookUrl": webhook_url,
            "metadata": metadata or {}
        }
        
        try:
            response = await self.client._make_request(
                "POST",
                "/api/executions",
                json=payload
            )
            
            return Execution.model_validate(response["data"])
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to start execution: {str(e)}")

    async def get_execution(self, execution_id: str) -> Execution:
        """
        Retrieve detailed information about a specific execution.
        
        Args:
            execution_id: ID of the execution to retrieve
            
        Returns:
            Complete execution object with all details
            
        Raises:
            ExecutionNotFoundError: If execution doesn't exist
            KlikkFlowAPIError: If API request fails
        """
        try:
            response = await self.client._make_request(
                "GET",
                f"/api/executions/{execution_id}"
            )
            
            return Execution.model_validate(response["data"])
            
        except Exception as e:
            if "404" in str(e):
                raise ExecutionNotFoundError(f"Execution {execution_id} not found")
            raise KlikkFlowAPIError(f"Failed to get execution: {str(e)}")

    async def list_executions(
        self,
        workflow_id: Optional[str] = None,
        status: Optional[ExecutionStatus] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        pagination: Optional[PaginationParams] = None
    ) -> PaginatedResponse[Execution]:
        """
        List executions with optional filtering and pagination.
        
        Args:
            workflow_id: Filter by specific workflow ID
            status: Filter by execution status
            start_date: Filter executions after this date
            end_date: Filter executions before this date
            pagination: Pagination parameters
            
        Returns:
            Paginated list of executions
        """
        params = {}
        
        if workflow_id:
            params["workflowId"] = workflow_id
        if status:
            params["status"] = status.value
        if start_date:
            params["startDate"] = start_date.isoformat()
        if end_date:
            params["endDate"] = end_date.isoformat()
            
        if pagination:
            params.update(pagination.model_dump(exclude_none=True))
        
        try:
            response = await self.client._make_request(
                "GET",
                "/api/executions",
                params=params
            )
            
            executions = [
                Execution.model_validate(exec_data) 
                for exec_data in response["data"]
            ]
            
            return PaginatedResponse(
                data=executions,
                total=response.get("total", len(executions)),
                page=response.get("page", 1),
                pageSize=response.get("pageSize", len(executions)),
                hasMore=response.get("hasMore", False)
            )
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to list executions: {str(e)}")

    async def cancel_execution(self, execution_id: str) -> bool:
        """
        Cancel a running execution.
        
        Args:
            execution_id: ID of the execution to cancel
            
        Returns:
            True if cancellation was successful
            
        Raises:
            ExecutionNotFoundError: If execution doesn't exist
            ExecutionError: If execution cannot be cancelled
        """
        try:
            response = await self.client._make_request(
                "POST",
                f"/api/executions/{execution_id}/cancel"
            )
            
            return response.get("success", False)
            
        except Exception as e:
            if "404" in str(e):
                raise ExecutionNotFoundError(f"Execution {execution_id} not found")
            if "cannot be cancelled" in str(e).lower():
                raise ExecutionError(f"Execution {execution_id} cannot be cancelled")
            raise KlikkFlowAPIError(f"Failed to cancel execution: {str(e)}")

    async def retry_execution(
        self, 
        execution_id: str,
        from_node: Optional[str] = None
    ) -> Execution:
        """
        Retry a failed execution, optionally from a specific node.
        
        Args:
            execution_id: ID of the failed execution to retry
            from_node: Optional node ID to retry from
            
        Returns:
            New execution object for the retry
        """
        payload = {}
        if from_node:
            payload["fromNode"] = from_node
            
        try:
            response = await self.client._make_request(
                "POST",
                f"/api/executions/{execution_id}/retry",
                json=payload
            )
            
            return Execution.model_validate(response["data"])
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to retry execution: {str(e)}")

    async def get_execution_logs(
        self, 
        execution_id: str,
        node_id: Optional[str] = None,
        level: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Retrieve logs for an execution, optionally filtered by node and level.
        
        Args:
            execution_id: ID of the execution
            node_id: Optional specific node ID
            level: Optional log level filter ('debug', 'info', 'warn', 'error')
            
        Returns:
            List of log entries
        """
        params = {}
        if node_id:
            params["nodeId"] = node_id
        if level:
            params["level"] = level
            
        try:
            response = await self.client._make_request(
                "GET",
                f"/api/executions/{execution_id}/logs",
                params=params
            )
            
            return response.get("data", [])
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to get execution logs: {str(e)}")

    async def stream_execution_updates(
        self, 
        execution_id: str
    ) -> AsyncGenerator[ExecutionData, None]:
        """
        Stream real-time updates for a running execution via WebSocket.
        
        Args:
            execution_id: ID of the execution to monitor
            
        Yields:
            ExecutionData objects with real-time updates
        """
        if not self._websocket_connection:
            await self._connect_websocket()
            
        # Subscribe to execution updates
        await self._websocket_connection.emit("subscribe", {
            "type": "execution",
            "executionId": execution_id
        })
        
        try:
            async for message in self._websocket_message_stream():
                if message.get("executionId") == execution_id:
                    yield ExecutionData.model_validate(message.get("data", {}))
                    
                    # Stop streaming if execution is complete
                    if message.get("data", {}).get("status") in ["completed", "failed", "cancelled"]:
                        break
                        
        finally:
            # Unsubscribe from execution updates
            if self._websocket_connection:
                await self._websocket_connection.emit("unsubscribe", {
                    "type": "execution",
                    "executionId": execution_id
                })

    async def get_node_execution_data(
        self, 
        execution_id: str, 
        node_id: str
    ) -> Optional[ExecutionNode]:
        """
        Get detailed execution data for a specific node.
        
        Args:
            execution_id: ID of the execution
            node_id: ID of the node
            
        Returns:
            ExecutionNode object with node execution details
        """
        try:
            response = await self.client._make_request(
                "GET",
                f"/api/executions/{execution_id}/nodes/{node_id}"
            )
            
            if response.get("data"):
                return ExecutionNode.model_validate(response["data"])
            return None
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to get node execution data: {str(e)}")

    def add_execution_listener(
        self, 
        execution_id: str, 
        callback: Callable[[ExecutionData], None]
    ) -> None:
        """
        Add a listener for execution updates.
        
        Args:
            execution_id: ID of the execution to listen to
            callback: Function to call when updates are received
        """
        if execution_id not in self._execution_listeners:
            self._execution_listeners[execution_id] = []
        
        self._execution_listeners[execution_id].append(callback)

    def remove_execution_listener(
        self, 
        execution_id: str, 
        callback: Callable[[ExecutionData], None]
    ) -> None:
        """
        Remove an execution listener.
        
        Args:
            execution_id: ID of the execution
            callback: Function to remove from listeners
        """
        if execution_id in self._execution_listeners:
            try:
                self._execution_listeners[execution_id].remove(callback)
            except ValueError:
                pass
            
            # Clean up empty listener lists
            if not self._execution_listeners[execution_id]:
                del self._execution_listeners[execution_id]

    async def _connect_websocket(self) -> None:
        """Connect to WebSocket for real-time updates."""
        # WebSocket connection implementation would go here
        # This would typically use a library like python-socketio
        pass

    async def _websocket_message_stream(self) -> AsyncGenerator[Dict[str, Any], None]:
        """Stream WebSocket messages."""
        # WebSocket message streaming implementation would go here
        # This is a placeholder for the actual implementation
        while True:
            await asyncio.sleep(1)  # Placeholder
            yield {}  # Placeholder

    async def export_execution_data(
        self, 
        execution_id: str, 
        format: str = "json"
    ) -> bytes:
        """
        Export execution data in various formats.
        
        Args:
            execution_id: ID of the execution to export
            format: Export format ('json', 'csv', 'xlsx')
            
        Returns:
            Exported data as bytes
        """
        try:
            response = await self.client._make_request(
                "GET",
                f"/api/executions/{execution_id}/export",
                params={"format": format}
            )
            
            # Convert response to appropriate format
            if format == "json":
                return json.dumps(response["data"]).encode()
            else:
                # For other formats, the API would return binary data
                return response.get("data", b"")
                
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to export execution data: {str(e)}")

    async def get_execution_statistics(
        self,
        workflow_id: Optional[str] = None,
        time_range: Optional[Dict[str, datetime]] = None
    ) -> Dict[str, Any]:
        """
        Get execution statistics and metrics.
        
        Args:
            workflow_id: Optional workflow ID to filter statistics
            time_range: Optional time range with 'start' and 'end' datetime objects
            
        Returns:
            Dictionary containing execution statistics
        """
        params = {}
        if workflow_id:
            params["workflowId"] = workflow_id
        if time_range:
            if "start" in time_range:
                params["startDate"] = time_range["start"].isoformat()
            if "end" in time_range:
                params["endDate"] = time_range["end"].isoformat()
                
        try:
            response = await self.client._make_request(
                "GET",
                "/api/executions/statistics",
                params=params
            )
            
            return response.get("data", {})
            
        except Exception as e:
            raise KlikkFlowAPIError(f"Failed to get execution statistics: {str(e)}")


# Convenience functions for common execution operations
async def start_workflow_execution(
    client: Any, 
    workflow_id: str, 
    input_data: Optional[Dict[str, Any]] = None
) -> Execution:
    """
    Convenience function to start a workflow execution.
    
    Args:
        client: KlikkFlowClient instance
        workflow_id: ID of the workflow to execute
        input_data: Optional input data for the workflow
        
    Returns:
        Execution object
    """
    execution_manager = ExecutionManager(client)
    return await execution_manager.start_execution(workflow_id, input_data)


async def wait_for_execution_completion(
    client: Any, 
    execution_id: str,
    timeout: Optional[int] = None
) -> Execution:
    """
    Wait for an execution to complete with optional timeout.
    
    Args:
        client: KlikkFlowClient instance
        execution_id: ID of the execution to wait for
        timeout: Optional timeout in seconds
        
    Returns:
        Final execution object
    """
    execution_manager = ExecutionManager(client)
    start_time = asyncio.get_event_loop().time()
    
    while True:
        execution = await execution_manager.get_execution(execution_id)
        
        if execution.status in [ExecutionStatus.COMPLETED, ExecutionStatus.FAILED, ExecutionStatus.CANCELLED]:
            return execution
            
        if timeout and (asyncio.get_event_loop().time() - start_time) > timeout:
            raise ExecutionError(f"Execution {execution_id} did not complete within {timeout} seconds")
            
        await asyncio.sleep(1)  # Poll every second